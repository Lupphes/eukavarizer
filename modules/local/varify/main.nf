process VARIFY {
    tag "$taxonomy_id"
    label 'process_low'
    cache false

    conda "${moduleDir}/environment.yml"
    container "docker.io/luppo/varify:ac8b977976d91caf2495d1e02cb135073fac722b"

    input:
    val taxonomy_id
    val outdir
    tuple val(meta1), path(survivor_vcf)
    tuple val(meta2), path(survivor_stats)
    tuple val(meta3), path(bcfmerge_vcf)
    tuple val(meta4), path(bcfmerge_tbi)
    tuple val(meta5), path(bcfmerge_stats)
    tuple val(meta6), path(vcf_list_cleaned)
    tuple val(meta7), path(reference_genome_bgzipped_index)
    val(profile)
    path(samtools_stats)

    output:
    path "*.html"       , emit: report_file
    path "plots/*.png"  , emit: report_images
    path "plots/*.html" , emit: report_html
    path "versions.yml" , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${taxonomy_id}"
    def samtools_stats_files = samtools_stats ? samtools_stats.collect { "--samtools-stats-file ${it}" }.join(' ') : ''
    """
    varify \\
        --output-dir . \\
        --bcf-vcf-file ${bcfmerge_vcf} \\
        --survivor-vcf-file ${survivor_vcf} \\
        --fasta-file ${reference_genome_bgzipped_index} \\
        --bcf-stats-file ${bcfmerge_stats} \\
        --survivor-stats-file ${survivor_stats} \\
        --report-file "${prefix}.html" \\
        --profile ${profile} \\
        # TODO: Add samtools parameter
        # ${samtools_stats_files} \\
        ${args}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        varify: \$( varify --version 2>&1 | sed 's/varify //g' )
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${taxonomy_id}"
    """
    mkdir -p plots
    touch ${prefix}.html
    touch plots/dummy.png
    touch plots/dummy.html

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        varify: \$( varify --version 2>&1 | sed 's/varify //g' )
    END_VERSIONS
    """
}
