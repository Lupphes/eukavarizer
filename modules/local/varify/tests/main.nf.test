nextflow_process {

    name "Test Process VARIFY"
    script "../main.nf"
    process "VARIFY"
    tag "varify"
    tag "modules"
    tag "modules_local"

    test("Generate report with samtools stats - stub") {
        options '-stub'

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = 4932  // taxonomy_id
                input[1] = "$outputDir"
                input[2] = [ [id: 'survivor'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf') ]
                input[3] = [ [id: 'survivor_stats'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta') ]
                input[4] = [ [id: 'bcfmerge'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz') ]
                input[5] = [ [id: 'bcfmerge_tbi'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz.tbi') ]
                input[6] = [ [id: 'bcfmerge_stats'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai') ]
                input[7] = [ [id: 'vcf_list'], [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test_annotate.vcf.gz') ] ]
                input[8] = [ [id: 'ref'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.dict') ]
                input[9] = "test"  // profile
                input[10] = [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.sizes') ]  // samtools_stats
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.report_file },
                { assert process.out.report_file.get(0).toString().endsWith('.html') },
                { assert process.out.versions }
            )
        }
    }

    test("Generate report without samtools stats - stub") {
        options '-stub'

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = 4932
                input[1] = "$outputDir"
                input[2] = [ [id: 'survivor'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf') ]
                input[3] = [ [id: 'survivor_stats'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta') ]
                input[4] = [ [id: 'bcfmerge'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz') ]
                input[5] = [ [id: 'bcfmerge_tbi'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz.tbi') ]
                input[6] = [ [id: 'bcfmerge_stats'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai') ]
                input[7] = [ [id: 'vcf_list'], [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test_annotate.vcf.gz') ] ]
                input[8] = [ [id: 'ref'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.dict') ]
                input[9] = "test"
                input[10] = []  // No samtools_stats
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.report_file },
                { assert process.out.report_images },
                { assert process.out.report_html },
                { assert process.out.versions }
            )
        }
    }

    test("Multiple samtools stats files - stub") {
        options '-stub'

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = 9606
                input[1] = "$outputDir"
                input[2] = [ [id: 'survivor'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf') ]
                input[3] = [ [id: 'survivor_stats'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta') ]
                input[4] = [ [id: 'bcfmerge'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz') ]
                input[5] = [ [id: 'bcfmerge_tbi'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz.tbi') ]
                input[6] = [ [id: 'bcfmerge_stats'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai') ]
                input[7] = [ [id: 'vcf_list'], [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test_annotate.vcf.gz') ] ]
                input[8] = [ [id: 'ref'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.dict') ]
                input[9] = "docker"
                input[10] = [
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.sizes'),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.bed')
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.versions }
            )
        }
    }
}
