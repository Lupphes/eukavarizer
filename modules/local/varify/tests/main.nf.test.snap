{
    "Multiple samtools stats files - stub": {
        "content": [
            {
                "0": [
                    "9606.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "1": [
                    "dummy.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "2": [
                    "dummy.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "3": [
                    "versions.yml:md5,255c4c63bec55e27f22c525d433efaf4"
                ],
                "report_file": [
                    "9606.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "report_html": [
                    "dummy.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "report_images": [
                    "dummy.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,255c4c63bec55e27f22c525d433efaf4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-20T22:37:30.744819548"
    },
    "Generate report with samtools stats - stub": {
        "content": [
            {
                "0": [
                    "4932.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "1": [
                    "dummy.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "2": [
                    "dummy.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "3": [
                    "versions.yml:md5,255c4c63bec55e27f22c525d433efaf4"
                ],
                "report_file": [
                    "4932.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "report_html": [
                    "dummy.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "report_images": [
                    "dummy.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,255c4c63bec55e27f22c525d433efaf4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-20T22:37:08.82982917"
    },
    "Generate report without samtools stats - stub": {
        "content": [
            {
                "0": [
                    "4932.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "1": [
                    "dummy.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "2": [
                    "dummy.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "3": [
                    "versions.yml:md5,255c4c63bec55e27f22c525d433efaf4"
                ],
                "report_file": [
                    "4932.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "report_html": [
                    "dummy.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "report_images": [
                    "dummy.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,255c4c63bec55e27f22c525d433efaf4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-20T22:37:19.230092226"
    }
}